### PR TITLE
feat(core): createPasscode

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

implement function: `createPasscode`

see LOG-846

## Testing

UTs

### CODE REVIEW

File: packages/core/src/lib/passcode.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  typeof deletePasscodeById
+>;
+const mockedInsertPasscode = insertPasscode as jest.MockedFunction<typeof insertPasscode>;
+
+beforeAll(() => {
+  mockedInsertPasscode.mockImplementation(async (data) => ({
+    ...data,
+    createdAt: Date.now(),
+    phone: data.phone ?? '',
+    email: data.email ?? '',
```
Comment (gao-sun) [inline packages/core/src/lib/passcode.test.ts:27]: no need for `??` since it's nullable?

File: packages/core/src/queries/passcode.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+import { buildInsertInto } from '@/database/insert-into';
+import pool from '@/database/pool';
+import { buildUpdateWhere } from '@/database/update-where';
+import { convertToIdentifiers } from '@/database/utils';
+import { DeletionError } from '@/errors/SlonikError';
+
+const { table, fields } = convertToIdentifiers(Passcodes);
+
+export const findUnconsumedPasscodeBySessionIdAndType = async (
```
Comment (gao-sun) [inline packages/core/src/queries/passcode.ts:12]: `.one()` throws an error if not found, also we should consider multiple unconsumed passcodes? maybe `.many()`?

Comment (gao-sun) [reply packages/core/src/queries/passcode.ts:12]: if so, we can delete those passcodes in batch (use pg `in` or `=any()` clause to test array)

File: packages/core/src/lib/passcode.test.ts (hunk 1 above)
Comment (wangsijie) [reply packages/core/src/lib/passcode.test.ts:27]: `data.phone` can be undefined, but the resolved `phone` cannot be undefined, so `??` is needed

File: packages/core/src/queries/passcode.ts (hunk 2 above)
Comment (wangsijie) [reply packages/core/src/queries/passcode.ts:12]: The number of unconsumed passcode for specific type and session is desigend to be 1 or 0.

But treat it as "multiple" would be more safer, updated.

File: packages/core/src/lib/passcode.test.ts (hunk 1 above)
Comment (gao-sun) [reply packages/core/src/lib/passcode.test.ts:28]: then should be `?? null`?

File: packages/core/src/lib/passcode.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export const createPasscode = async (
+  sessionId: string,
+  type: PasscodeType,
+  payload: { phone: string } | { email: string }
+) => {
+  // Disable existing passcodes.
+  const passcodes = await findUnconsumedPasscodesBySessionIdAndType(sessionId, type);
+  if (passcodes.length > 0) {
+    await deletePasscodesByIds(passcodes.map((passcode) => passcode.id));
```
Comment (gao-sun) [inline packages/core/src/lib/passcode.ts:20]: nit:
```suggestion
    await deletePasscodesByIds(passcodes.map(({ id }) => id));
```

**Review by gao-sun — APPROVED** [review]
lgtm

File: packages/core/src/lib/passcode.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { PasscodeType } from '@logto/schemas';
+import { customAlphabet, nanoid } from 'nanoid';
+
+import {
+  deletePasscodesByIds,
+  findUnconsumedPasscodesBySessionIdAndType,
+  insertPasscode,
+} from '@/queries/passcode';
+
+const randomCode = customAlphabet('1234567890', 6);
```
Comment (darcyYe) [inline packages/core/src/lib/passcode.ts:10]: should we make the length of the randomCode / passcode configurable? or provide a 4-digit random code generator as well?

File: packages/core/src/queries/passcode.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
```
Comment (darcyYe) [inline packages/core/src/queries/passcode.ts:12]: this method is unused, should we keep it?

**Review by darcyYe — CHANGES_REQUESTED** [review]

File: packages/core/src/lib/passcode.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+export const createPasscode = async (
+  sessionId: string,
+  type: PasscodeType,
+  payload: { phone: string } | { email: string }
+) => {
+  // Disable existing passcodes.
+  const passcodes = await findUnconsumedPasscodesBySessionIdAndType(sessionId, type);
```
Comment (simeng-li) [inline packages/core/src/lib/passcode.ts:18]: nit: not 100% sure, but, do you think we could make the query method more flexible? Like 

```
findPasscodes({
 sessionId,
 type,
 consumed: false
})
```

File: packages/core/src/queries/passcode.ts (hunk 5 above)
Comment (wangsijie) [reply packages/core/src/queries/passcode.ts:12]: keep if for now, would be used in `verifyPasscode`

File: packages/core/src/lib/passcode.ts (hunk 4 above)
Comment (wangsijie) [reply packages/core/src/lib/passcode.ts]: Change into a constant.

**Review by darcyYe — APPROVED** [review]
Lgtm

**Pull request merged**